### ISSUE TITLE
Doc of mutex advicing std blocking mutex look totally misleading

### ISSUE THREAD

**Issue by Stargateur** [tokio-rs/tokio#5024]
[Doc](https://docs.rs/tokio/latest/tokio/sync/struct.Mutex.html#which-kind-of-mutex-should-you-use) of mutex currently state:

> Contrary to popular belief, it is ok and often preferred to use the ordinary [Mutex](https://doc.rust-lang.org/nightly/std/sync/mutex/struct.Mutex.html) from the standard library in asynchronous code.

That contrary to all my understanding and all the doc of async I know. Tokio doc is very clear about a task **should not block**. Waiting on a std mutex is blocking, I don't see how the doc can suggest to use std mutex.

**Comment by Darksonn** [comment]
Whether or not something counts as blocking has to do with the duration it takes. Usually, you will be blocking on the mutex for a sufficiently short time that you avoid problems.

**Comment by Stargateur** [comment]
> Whether or not something counts as blocking has to do with the duration it takes.

First time I read this, thus that mostly true in practice in theory that false, I think that asking for trouble. I think use mutex (as a user) in async is often a mistake, you better use channel and co.

I have trouble to understand exactly the doc:

> The feature that the async mutex offers over the blocking mutex is the ability to keep it locked across an .await point.

> Note that, although the compiler will not prevent the **std** Mutex from holding its guard across .await points in situations where the task is not movable between threads, this virtually never leads to correct concurrent code in practice as it can easily lead to deadlocks.

The doc mix std and tokio mutex implementation and is hard to follow, and I don't understand how tokio mutex would not deadlock like std mutex since the doc say "keep it locked across an .await point." for both mutex.

It's look like the doc should be separate in 3 sections:

- don't use this
- how to use std mutex in async
- how to use tokio mutex

**Comment by Darksonn** [comment]
I'm sure this particular section in the documentation could be improved, and I have a draft for a article that discusses how the std mutex is used in more detail, but I haven't had the time to work on it for a while.

The Tokio mutex doesn't deadlock when held across an `.await` because the `lock` method uses an `.await`, which allows the task to yield to the runtime while waiting for the mutex to become available.

Yes, it also applies to `RwLock`.

**Comment by Darksonn** [comment]
To make my point clearer: Yes, there exist extremely simple programs where holding an std mutex locked over an await does not deadlock. Typically that is because you never have two tasks call `lock` at the same time. But such cases are very rare in real-world programs, and in essentially all real-world situations, holding an std mutex locked over an await is incorrect.

**Comment by jobafr** [comment]
So, as I understand it now, that first sentence ("The feature that the async mutex offers over the blocking mutex is the ability to keep it locked across an .await point.") could be rephrased as "The feature that the async mutex offers over the blocking mutex is the ability to be locked over an await point without deadlocking the executor on concurrent attempts to lock the same mutex".

Now, since the docs also say "is ok and often preferred to use the ordinary [Mutex](https://doc.rust-lang.org/nightly/std/sync/mutex/struct.Mutex.html) from the standard library in asynchronous code", I wonder what the "correct" use case for std's mutex in async code is. Sharing something between sync code and async code?

**Comment by tbodt** [comment]
https://github.com/tokio-rs/tokio/pull/6819

**Comment by jokeyrhyme** [comment]
whilst we're reviewing the recommendations around Mutex, oxide.computer now recommends that their own staff completely avoid tokio's Mutex as it is a common cause of cancellation bugs for them

https://youtu.be/zrv5Cy1R7r4 "cancelling async Rust"

https://rfd.shared.oxide.computer/rfd/0609 "FutureLock"

is it worth more broadly deprecating tokio's Mutex until the debugging and cancellation difficulties are reduced?

**Comment by Darksonn** [comment]
The problem with FutureLock is `select!`, not `Mutex`.

**Comment by jokeyrhyme** [comment]
@Darksonn not quite, RE: `select!` and `Mutex`:

> None of these by itself is wrong, but combining them results in futurelock. Remember too that the Mutex could be buried beneath several layers of function calls in different modules or packages. It could require looking across many layers of the stack at once to be able to see the problem.

that RFD i linked refers to https://rfd.shared.oxide.computer/rfd/397 and https://rfd.shared.oxide.computer/rfd/400 which are much more on-topic

the general issue here is "cancellation correctness" as explained at https://rfd.shared.oxide.computer/rfd/400#_cancel_correctness

> Cancel correctness is violated if all three of the following conditions are met:
> 
> 1. The system has a cancel-unsafe future somewhere in it.
> 2. This future is cancelled.
> 3. Cancelling the future violates some property of the system.

the issue with tokio's Mutex, as explained at https://rfd.shared.oxide.computer/rfd/397#_example_with_mutexes , is that it's deceptively simple to adopt as an alternative to std::sync::Mutex , this makes it a highly-frequent contributor to the violation of cancellation correctness (due to being cancel-unsafe)

**Comment by jokeyrhyme** [comment]
use of cancel-unsafe Futures probably needs as much focused and expert review as `unsafe` blocks, perhaps even more so because the entire call-chain needs to be considered with extreme caution

and this isn't a tokio-specific issue, i'm only suggesting that anywhere tokio provides a cancel-unsafe future (e.g. Mutex) that there be much more warning than what we currently have as of 1.48.0: https://docs.rs/tokio/latest/tokio/sync/struct.Mutex.html#cancel-safety

> Cancel safety
>
> This method uses a queue to fairly distribute locks in the order they were requested. Cancelling a call to lock makes you lose your place in the queue.

depending on the behaviour of the entire program, the side-effect of cancellation here may trigger all 3 factors required to violate cancellation correctness, and the only way to know for sure (so far) is to have a human carefully review the entire call chain

could we deprecate Mutex (and other cancel-unsafe futures) or otherwise improve the documentation and/or tooling (debugging, linters, etc) so that this isn't a footgun?

**Comment by Darksonn** [comment]
I mean, ultimately cancel safety something we have battled for years. Tokio's mutex is just one small example, and it sounds like the cases you are thinking of are not cases where `Mutex::lock()` is cancelled, but rather other cases where you cancel some generic async fn containing a `MutexGuard`.

I think the traditional view is that mechanisms for cancellation such as `tokio::select!` need to be much more limited, and this is how e.g. [`futures-concurrency`](https://docs.rs/futures-concurrency) came into the picture. But you're right that in any situation where combining A+B is problematic, then saying "A is the problem" is an incomplete view of the situation.
